libs/ui/src/utils/dapp.utils.ts

```diff
@@ -58,10 +39,34 @@ const createDAppNotificationResponse = <T>(method: string, params: T) => ({
   target: 'oko-inpage'
 });
 
-export const sendNotificationToDApp = <T>(method: string, result: T) => {
-  tabs.query({ active: true }).then(queryTabs => {
-    if (queryTabs[0].id !== undefined) {
-      tabs.sendMessage(queryTabs[0].id, createDAppNotificationResponse(method, result));
+const sendMessageToDAppTab = (response: unknown, origin: string) => {
+  tabs.query({}).then(queryTabs => {
+    const dAppTab = queryTabs.find(tab => tab.url?.includes(origin));
+    if (isDefined(dAppTab) && isDefined(dAppTab.id)) {
+      tabs.sendMessage(dAppTab.id, response);
     }
   });
 };
+
+export const sendResponseToDAppAndClosePopup = <T>(id: string, result: T, origin: string) => {
+  const response = createDAppResponse(id, result);
+  sendMessageToDAppTab(response, origin);
+
+  setTimeout(() => close(), 1000);
+};
+
+export const sendErrorToDAppAndClosePopup = (id: string, origin: string) => {
+  const errorResponse = createErrorMessage(id);
+  sendMessageToDAppTab(errorResponse, origin);
+
+  setTimeout(() => close(), 1000);
+};
+
+export const sendNotificationToDApp = <T>(method: string, result: T, origin: string) => {
+  const notification = createDAppNotificationResponse(method, result);
+  sendMessageToDAppTab(notification, origin);
+};
+
+export const sendMessageToBackground = () => {
+  runtime.sendMessage(undefined, { type: 'POPUP_CLOSED' });
```

Comment:
move POPUP_CLOSED & POPUP_OPEN into libs/ui and reexport from libs/ui/src/background-script